File: solutions/Path-Sum-II/Path-Sum-II.cpp

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void pathSumHelper(TreeNode *root,int sum,vector<vector<int> > &ans,vector<int> vec,int cur)
    {
        if(sum==cur && !root->left && !root->right)
        {
            ans.push_back(vec);
            return ;
        }
        if(root->left)
        {
            vec.push_back(root->left->val);
            pathSumHelper(root->left,sum,ans,vec,cur+root->left->val);
            vec.pop_back();
        }
        if(root->right)
        {
            vec.push_back(root->right->val);
            pathSumHelper(root->right,sum,ans,vec,cur+root->right->val);
            vec.pop_back();
        }
    }
    vector<vector<int> > pathSum(TreeNode *root, int sum) {
        vector<vector<int> >ans;
        vector<int>vec;
        if(!root)
            return ans;
        vec.push_back(root->val);
        pathSumHelper(root,sum,ans,vec,root->val);
        return ans;
    }
};


 

File: solutions/Path-Sum-II/Path-Sum-II.py

‘’’
 Definition for a  binary tree node
 class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None
‘’’
class Solution:
    def dfs(self,root,sum,cnt,vec,ans):
        if sum==cnt and root.left==None and root.right==None:
            vec2=vec[:]
            ans.append(vec2)
            return
        if root.left!=None:
            vec.append(root.left.val)
            self.dfs(root.left,sum,cnt+root.left.val,vec,ans)
            vec.pop()
        if root.right!=None:
            vec.append(root.right.val)
            self.dfs(root.right,sum,cnt+root.right.val,vec,ans)
            vec.pop()
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        ans=[]
        vec=[]
        if root==None:
            return ans
        vec.append(root.val)
        self.dfs(root,sum,root.val,vec,ans)
        return ans
        
 